### PR TITLE
Script to abort outdated PR builds

### PR DESCRIPTION
**Pull request by clytwynec**
@jzoldak @benpatterson Please review.

Requires a Jenkins user and api key along with the URL for the ghprb job.
```
python ghprb_deduper.py  -t JENKINS_API_KEY -u JENKINS_USER -j JOB_URL
```

Console output looks like:
```
[INFO] Updating description for build #71. Response: 200
[INFO] Aborting build #71. Response: 200
[INFO] 
---------------------------------
** Extra running builds found. **
---------------------------------
PR #11:
	Num running builds: 2
	Current build: 72
	Outdated builds: [71]
```
OR
```
[INFO] No extra running builds found.
```

The updated description on the aborted build looks like:
```
[PR ​#11] ​​Build ​​automatically ​​aborted ​​because there ​​is ​​a ​​newer ​​build ​​for ​​the ​​same ​​PR. ​​See ​​build ​​#72.
```

### CODE REVIEW

File: scripts/ghprb_deduper.py [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+    requests==2.5.3
+"""
+from collections import defaultdict
+from operator import itemgetter
+import argparse
+import json
+import re
+import requests
+import logging
+logging.basicConfig(format='[%(levelname)s] %(message)s')
```
Comment (benpatterson) [inline scripts/ghprb_deduper.py]: nit: insert an empty line

File: scripts/ghprb_deduper.py [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+    def update_build_desc(self, build_id, description):
+        """
+        Updates build description.
+
+        :Args:
+            build_id: id number of build to update
+            description: the new description
+        """
+        build_url = self._append_url(self.job_url, str(build_id))
+        url = self._append_url(build_url, "/submitDescription")
```
Comment (benpatterson) [inline scripts/ghprb_deduper.py]: We'll need to make sure the user that makes these edits has this level of access.

File: scripts/ghprb_deduper.py [hunk 3]
```diff
@@ -0,0 +1,237 @@
+"""
```
Comment (benpatterson) [inline scripts/ghprb_deduper.py]: Nit...insert a description of what this script does.

File: scripts/ghprb_deduper.py [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+
+    def stop_build(self, build_id):
+        """
+        Stops a build.
+
+        :Args:
+            build_id: id number of build to abort
+        """
+        build_url = self._append_url(self.job_url, str(build_id))
+        url = self._append_url(build_url, "/stop")
```
Comment (benpatterson) [inline scripts/ghprb_deduper.py]: same here w.r.t. perms, of course

**Comment by benpatterson** [comment]
Have time to add a couple of tests? Particularly stop_duplicate_builds, and in general, making sure we can handle multiple PRs (some with dupes, some without dupes). Not a must, but a nice-to-do.

Reading through the code, LGTM.

File: scripts/ghprb_deduper.py [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+                ** building: Boolean of whether it is actively building
+                ** timestamp: the time the build started
+                ** number: the build id number
+                ** actions: a list of 'actions', from which the only
+                    item used in this script is 'parameters' which can
+                    be used to find the PR number.
+        """
+        api_url = self._append_url(self.job_url, '/api/json')
+
+        response = requests.get(
```
Comment (jzoldak) [inline scripts/ghprb_deduper.py]: Should this be in a try/catch block? I'm wondering what the timeout is and what will happen if jenkins master takes too long to respond.

File: scripts/ghprb_deduper.py [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+                lines.append("\tCurrent build: {}".format(
+                    current_build_id))
+                lines.append("\tOutdated builds: {}".format(
+                    old_build_ids))
+
+                desc = self._aborted_description(current_build_id, pr)
+
+                for b in old_build_ids:
+                    try:
+                        self.update_build_desc(b, desc)
```
Comment (jzoldak) [inline scripts/ghprb_deduper.py]: If the build finished up while we were calculating which ones to abort, then this will update the build desciption saying that we stopped it when we didn't. That might be confusing.
I'm imagining that you can update the build description after stopping it tho, so maybe switch these two line and if the aborting of the build is caught then the desc won't be updated.
This is assuming that if you call stop and the build is no longer running, it will get caught as an exception. Not sure if that's true.

Comment (clytwynec) [inline scripts/ghprb_deduper.py]: Good point,  you can update the description after a build has stopped.  I don't think it will raise an exception if the build is already finished, but I think the response should have an indication of whether or not it actually aborted.

**Comment by jzoldak** [comment]
The code looks nice. 

+1 on tests tho. I imagine this functionality is going to be so useful to us that we are going to want to extend it, so we should put the test framework in place now. Also if there are any weird/corner case bugs that crop up we'll spend a lot of time chasing them down if we don't proactively squash them.

I'd like to see all public methods tested, and also failure conditions of invalid credentials and job name.

File: scripts/ghprb_deduper.py (hunk 5 above)
Comment (clytwynec) [inline scripts/ghprb_deduper.py]: As written, the rest of the script will just fail because there is no data to parse.  Same for the case where a wrong url is passed. If it is a wrong url that happens to be using ghprb and the user has permissions to it, then it will all work...  It's relying on the user to input correct things. 

I would probably opt for using `response.raise_for_status()` instead of a try/except here.

**Comment by clytwynec** [comment]
Does the structure of this repo make sense?  I'm mainly thinking about having a clear way of locating requirements and tests.  It seems like it may make more sense to organize it by functionality, rather than `scripts` and `other stuff`.

**Comment by benpatterson** [comment]
In terms of the repo org....yes that's a good point. It's a bit early right now to figure out what will make the most sense, but maybe we put a stake in the ground that will make things more obvious and force us to rethink structure next time. 

Like, 

<root>/duplicate-running-jobs/script.py

In terms of high-level stuff, I noticed at this repo is actually a fork from your account. Would it make more sense to have just moved your repo to edx?

**Comment by clytwynec** [comment]
Probably... Does it make sense to go back an do that now?

**Comment by benpatterson** [comment]
@clytwynec Just occurred to me that we may need to re-open this PR. So, it might be easier to get this work done, then pull it upstream under your account...and at that point when all is equal, do the switcheroo.

File: scripts/ghprb_deduper.py (hunk 5 above)
Comment (jzoldak) [inline scripts/ghprb_deduper.py]: Cool. If you've figured out what happens in error situations, I would recommend that you code them to give a specific response to make troubleshooting failures easier.
Another note is that since we are indexing jenkins build data with splunk we'll be able to query on stuff, which is great.

**Comment by clytwynec** [comment]
Ok.  @benpatterson @jzoldak  I finally got around to catching up on this.    The functionality for the script remains the same, with the couple of earlier comments addressed.  The main changes in the latest commit:
* added tests
* requirements files (for both tests and for the script)
* a .gitignore file
* a travis file, so the tests (with coverage) and pep8 ran on travis.

File: duplicate_ghprb_jobs/deduper.py [hunk 7]
```diff
@@ -0,0 +1,252 @@
+"""
+This script is intended to be used to abort outdated builds
+on a jenkins job that uses the ghprbPlugin. In this case,
+an outdate build is one for which the PR that triggered it
```
Comment (benpatterson) [inline duplicate_ghprb_jobs/deduper.py]: nit: s/outdate/outdated

File: duplicate_ghprb_jobs/tests/test_deduper.py [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+        ('duplicate_ghprb_jobs.deduper.GhprbOutdatedBuildAborter.'
+         '_aborted_description'),
+        return_value='new description')
+    def test_stop_duplicates_with_duplicates(self, mock_desc,
+                                             update_desc,
+                                             stop_build):
+        sample_buid_data = sample_data(['1', '2', '2', '3'], [])
+        build_data = self.deduper.get_running_builds(sample_buid_data)
+        response = self.deduper.stop_duplicates(build_data)
+        stop_build.assert_called_once_with(1)
```
Comment (benpatterson) [inline duplicate_ghprb_jobs/tests/test_deduper.py]: should: give a bit more detail (for example, 'out of the 4 builds, there is one duplicate (a "2"), so this call should only be made for that duplicate')

File: duplicate_ghprb_jobs/tests/test_deduper.py [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+        update_desc.assert_called_once_with(1, mock_desc())
+
+    @patch(
+        'duplicate_ghprb_jobs.deduper.GhprbOutdatedBuildAborter.stop_build',
+        side_effect=HTTPError())
+    @patch(
```
Comment (benpatterson) [inline duplicate_ghprb_jobs/tests/test_deduper.py]: I actually didn't realize the `@patch` decorator could be used multiple times on one method. Pro tip!

**Comment by benpatterson** [comment]
Thanks for adding tests. A couple updates could help but no musts from me. :+1:

**Comment by jzoldak** [comment]
LGTM :+1:
For the next iteration on this (if we feel the need), I think we should tweak/improve the console output so that we can more easily get stats via splunk on how many builds were aborted, etc.
